stdlib/LibGit2/src/callbacks.jl

```diff
@@ -359,19 +359,131 @@ function fetchhead_foreach_callback(ref_name::Cstring, remote_url::Cstring,
+        count = @ccall libssh2_knownhost_readfile(
+            hosts :: Ptr{Cvoid},
+            file  :: Cstring,
+            1     :: Cint, # standard OpenSSH format
+        ) :: Cint
+        if count < 0
+            @warn("Error parsing SSH known hosts file `$file`")
+            @ccall libssh2_knownhost_free(hosts::Ptr{Cvoid})::Cvoid
+            continue
+        end
+        # We can't use libssh2_knownhost_check the normal way because libgit2,
+        # for no good reason, doesn't give us the fingerprint that we can use
+        # for that and instead gives us multiple hashes of that fingerprint.
+        # Instead, we have to look up the host with an incorrect fingerprint and
+        # then compute hashes of that to compare to what the callback gets from
+        # libgit2's pointlessly unhelpful API. Y U NO GIVE FINGERPRINT?
+        type = Consts.LIBSSH2_KNOWNHOST_TYPE_PLAIN |
+               Consts.LIBSSH2_KNOWNHOST_KEYENC_BASE64
+        hostr = Ref{Ptr{KnownHost}}()
+        check = @ccall libssh2_knownhost_check(
+            hosts :: Ptr{Cvoid},
+            host  :: Cstring,
+            ""    :: Ptr{UInt8},
+            0     :: Csize_t,
+            type  :: Cint,
+            hostr :: Ptr{Ptr{KnownHost}},
+        ) :: Cint
+        if check == Consts.LIBSSH2_KNOWNHOST_CHECK_NOTFOUND
+            # nothing relevant, try the next file
+            @ccall libssh2_knownhost_free(hosts::Ptr{Cvoid})::Cvoid
+            continue
+        elseif check == Consts.LIBSSH2_KNOWNHOST_CHECK_FAILURE
+            @warn("Error searching SSH known hosts file `$file`")
+            @ccall libssh2_knownhost_free(hosts::Ptr{Cvoid})::Cvoid
+            continue
+        end
+        @assert check == Consts.LIBSSH2_KNOWNHOST_CHECK_MISMATCH
+        # got a known hosts record for host, now check its key hash
+        key = base64decode(unsafe_string(unsafe_load(hostr[]).key))
+        match = true # avaliable hashes must match & there's at least one
+        if key_hashes.mask & Consts.CERT_SSH_SHA1 != 0
+            match &= sha1(key) == collect(key_hashes.sha1)
+        end
+        if key_hashes.mask & Consts.CERT_SSH_SHA256 != 0
+            match &= sha256(key) == collect(key_hashes.sha256)
+        end
+        match || verify_host_error("SSH host verification: the identity of the server `$host` does not match its known hosts record. Someone could be trying to man-in-the-middle your connection. It is also possible that the server has changed its key, in which case you should check with the server administrator and use `ssh $host` to login into the server and update your known hosts file.")
+        @ccall libssh2_knownhost_free(hosts::Ptr{Cvoid})::Cvoid
+        @assert 0 == @ccall libssh2_session_free(session::Ptr{Cvoid})::Cint
+        return match
+    end
+    verify_host_error("SSH host verification: the server `$host` is not a known host. Please connect once using `ssh $host` in order to add the server to your known hosts file.")
```

Comment:
Suggest `ssh-keyscan -H $host >> $file` as an alternative? I'm worried people will be confused by the ssh connection erroring (since git hosts tend to not have shell access).

Reply:
Want to make a suggestion with the appropriate command syntax?

Reply:
I'm a bit worried about two things:

1. Someone not having `ssh-keyscan` for example if they are on Windows, using Putty.
2. Someone accidentally clobbering or corrupting their `~/.ssh/known_hosts` file this way.

Since the `ssh` command will automatically, and presumably safely, add keys to the known hosts file for you, that seems better. If we're not worried about the corruption possibility, then we could use `Sys.which("ssh-keyscan")` to check if the `ssh-keyscan` command exists or not and print a suggesting that is conditional on that. I can make a potential change along these lines.

Reply:
I've added this: it checks `Sys.which("ssh-keyscan")` to decide what to tell the user to try.